lib/fluent/plugin_helper/compat_parameters.rb

```diff
@@ -40,59 +41,185 @@ module CompatParameters
-            next unless current
-            if conf.has_key?(compat)
-              if compat == 'buffer_queue_full_action' && conf[compat] == 'exception'
-                attr[current] = 'throw_exception'
-              else
-                attr[current] = conf[compat]
-              end
-            end
+      def compat_parameters_buffer(conf, default_chunk_key: '')
+        # return immediately if <buffer> section exists, or any buffer-related parameters don't exist
+        return unless conf.elements('buffer').empty?
+        return if (BUFFER_PARAMS.keys + BUFFER_TIME_SLICED_PARAMS.keys).all?{|k| !conf.has_key?(k) }
+
+        # TODO: warn obsolete parameters if these are deprecated
+        buffer_params = BUFFER_PARAMS.merge(BUFFER_TIME_SLICED_PARAMS)
+        attr = compat_parameters_copy_to_subsection_attributes(conf, buffer_params) do |compat_key, value|
+          if compat_key == 'buffer_queue_full_action' && value == 'exception'
+            'throw_exception'
+          else
+            value
           end
-          TIME_SLICED_PARAMS.each do |compat, current|
-            next unless current
-            attr[current] = conf[compat] if conf.has_key?(compat)
+        end
+
+        chunk_key = default_chunk_key
+
+        if conf.has_key?('time_slice_format')
+          chunk_key = 'time'
+          attr['timekey'] = case conf['time_slice_format']
+                            when /\%S/ then 1
+                            when /\%M/ then 60
+                            when /\%H/ then 3600
+                            when /\%d/ then 86400
+                            else
+                              raise Fluent::ConfigError, "time_slice_format only with %Y or %m is too long"
+                            end
+        else
+          if chunk_key == 'time'
+            attr['timekey'] = 86400 # TimeSliceOutput.time_slice_format default value is '%Y%m%d'
           end
+        end
 
-          chunk_key = nil
-
-          if conf.has_key?('time_slice_format')
-            chunk_key = 'time'
-            attr['timekey'] = case conf['time_slice_format']
-                              when /\%S/ then 1
-                              when /\%M/ then 60
-                              when /\%H/ then 3600
-                              when /\%d/ then 86400
-                              else
-                                raise Fluent::ConfigError, "time_slice_format only with %Y or %m is too long"
-                              end
+        e = Fluent::Config::Element.new('buffer', chunk_key, attr, [])
+        conf.elements << e
+
+        conf
+      end
+
+      def compat_parameters_inject(conf)
+        return unless conf.elements('inject').empty?
+        return if INJECT_PARAMS.keys.all?{|k| !conf.has_key?(k) }
+
+        # TODO: warn obsolete parameters if these are deprecated
+        attr = compat_parameters_copy_to_subsection_attributes(conf, INJECT_PARAMS) do |compat_key, value|
```

Comment:
Ah, it's not needed... I think it remains after some trial&errors. I'll remove it.